base_serializer.go

```diff
@@ -24,48 +30,155 @@ type baseSerializer struct {
 	hasContainerAnnotation bool
 }
 
-func (baseSerializer *baseSerializer) stepIn(ionValue interface{}) {
-	panic("implement me")
+func (baseSerializer *baseSerializer) stepIn(ionValue hashValue) error {
+	err := baseSerializer.handleFieldName(ionValue)
+	if err != nil {
+		return err
+	}
+
+	err = baseSerializer.handleAnnotationsBegin(ionValue, true)
+	if err != nil {
+		return err
+	}
+
+	baseSerializer.beginMarker()
+
+	tq := tq(ionValue)
+	if ionValue.isNull() {
+		tq = tq | 0x0F
+	}
+
+	baseSerializer.update([]byte{tq})
+
+	return nil
 }
 
-func (baseSerializer *baseSerializer) digest() []byte {
-	panic("implement me")
+func (baseSerializer *baseSerializer) Sum(b []byte) []byte {
+	return baseSerializer.hashFunction.Sum(b)
 }
 
-func (baseSerializer *baseSerializer) handleFieldName(ionValue interface{}) {
-	panic("implement me")
+func (baseSerializer *baseSerializer) handleFieldName(ionValue hashValue) error {
+	if baseSerializer.depth > 0 && ionValue.isInStruct() {
+		// SymbolTokens are not available right now.
+		/*if ionValue.fieldNameSymbol().Text == null && ionValue.fieldNameSymbol().Sid != 0 {
+			return &UnknownSymbolError{ionValue.fieldNameSymbol().Sid}
+		}
+		return baseSerializer.writeSymbol(ionValue.fieldNameSymbol().Text)*/
+	}
+
+	return nil
 }
 
 func (baseSerializer *baseSerializer) update(bytes []byte) {
-	panic("implement me")
+	baseSerializer.hashFunction.Update(bytes)
 }
 
 func (baseSerializer *baseSerializer) beginMarker() {
-	panic("implement me")
+	baseSerializer.hashFunction.Update([]byte{BeginMarkerByte})
 }
 
 func (baseSerializer *baseSerializer) endMarker() {
-	panic("implement me")
+	baseSerializer.hashFunction.Update([]byte{EndMarkerByte})
 }
 
-func (baseSerializer *baseSerializer) handleAnnotationsBegin(ionValue interface{}, isContainer bool) {
-	panic("implement me")
+func (baseSerializer *baseSerializer) handleAnnotationsBegin(ionValue hashValue, isContainer bool) error {
+	annotations := ionValue.getAnnotations()
+	if len(annotations) > 0 {
+		baseSerializer.beginMarker()
+		baseSerializer.update([]byte{TqValue})
+		for _, annotation := range annotations {
+			err := baseSerializer.writeSymbol(annotation)
+			if err != nil {
+				return err
+			}
+		}
+
+		if isContainer {
+			baseSerializer.hasContainerAnnotation = true
+		}
+	}
+
+	return nil
 }
 
-func (baseSerializer *baseSerializer) handleAnnotationsEnd(ionValue interface{}, isContainer bool) {
-	panic("implement me")
+func (baseSerializer *baseSerializer) handleAnnotationsEnd(ionValue hashValue, isContainer bool) {
+	if (ionValue != nil && len(ionValue.getAnnotations()) > 0) ||
+		(isContainer && baseSerializer.hasContainerAnnotation) {
+		baseSerializer.endMarker()
+
+		if isContainer {
+			baseSerializer.hasContainerAnnotation = false
+		}
+	}
 }
```

Comment:
Currently the code behaves the same if we have annotations on a value that is scalar or container. So I do not see the reason for keeping 2 booleans (one as argument and on in the serializer). Is there some other reason that is yet to be implemented that treats annotations of containers and scalars differently?

Reply:
I don't see a reason to keeping the flag either. @justing-bq Lets remove it. Since this was ported from dotnet, we can create an issue to remove it on dotnet side aswell.

Reply:
Removed.